Fix using ovn kubernetes as the CNI

astoycos opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

This PR re-writes the submariner networkplugin-syncer for ovn-kubernetes.   Mainly it converts from using go-ovn/nbctl to libovsdb for interacting with the ovn-dbs. I will update this PR with more commits/explanations as I finish tidying things up. 

Fixes https://github.com/submariner-io/submariner/issues/1608

Depends on https://github.com/ovn-org/ovn-kubernetes/pull/2747
Depends on https://github.com/submariner-io/shipyard/pull/704

astoycos committed:
Add Constants to a single file

Around the ovn package we defined constants in a
confusing array of locations, this commit reorganizes
them into a single file.

Signed-off-by: Andrew Stoycos <astoycos@redhat.com>

astoycos committed:
Remove nbctl pkg

With the conversion to using libovsdb we no longer need to
use nbctl. This commit removes the associated helper
functions.

Signed-off-by: Andrew Stoycos <astoycos@redhat.com>

astoycos committed:
Move core-infra functionality and use libovsdb

Rename the files which contain the core submariner/ovn
functionality bits. Convert all the ovsdb clients from
go-ovn to ovsdb.

Re-organize some submariner files to align more
closely with OVN Semantics.  Convert all calls
reguarding OVN logical_router_static_routes to
use libovsdb rather than go-ovn.

Convert all logical_router_policy calls
to libovsdb.

convert all OVN sbdb chassis
related calls to use libovsdb.

Update how clusterSet service avaliability works
in the networkplugin-syncer code. Speciafically it ensures we add
the a new OVN object (OVN cluster loadbalancer_group) to the submariner
router. This group has all the VIPs for the cluster in question.

Update vendoring since ovn-k requires k8s 1.23

Signed-off-by: Andrew Stoycos <astoycos@redhat.com>

astoycos commented:
I would expect E2E to fail until the dependent PRs (updating our KIND setup) works

astoycos commented:
I've tested this locally by running E2E and things look good so far, so I'll go ahead and open this up for review.

mangelajo commented:
Very cool, @astoycos pong me to see if I can do a review. I'm not sure if I have enough bandwidth currently to go through the whole patch / do a detailed review, I will try do diagonally-read as much as possible.

I'd make sure it works with OCP + kind.

astoycos committed:
Setup + initialize libovsdb Clients

Switch from using go-ovn clients to interact
with the ovn dbs and change to using libovsdb.

Signed-off-by: Andrew Stoycos <astoycos@redhat.com>

astoycos commented:
@skitt @dfarrell07 Currently in the `End to End Full / E2E (operator, vxlan, ovn) (pull_request)` workflow... a major false positive considering ovn-k isn't running at all

```
[14:35:21.497] [submariner]$ [cluster2] cut -f 2- -d=
Cluster settings::
  broker - 'cluster1'
  clusters - 'cluster1' 'cluster2'
  cni - ([cluster2]="weave" [cluster1]="weave" )
  nodes per cluster - ([cluster2]="control-plane worker worker" [cluster1]="control-plane worker" )
```

skitt commented:
The dependencies have been merged, @astoycos this should be good to go once you rebase it!

astoycos committed:
Ensure we cleanup cableengine resources

With the OVN-K CNI we were seeing and ungraceful
termination of the GW Pod in GW migration tests,
this patch ensures we cleanup the cableengine
on node resources before exiting.

Signed-off-by: Andrew Stoycos <astoycos@redhat.com>

astoycos commented:
```
    nginx-demo:
      Container ID:  containerd://5f33ec7720d53392dc1ffbaf76c83df53a293e41aa9dba5ec0a2f0fe32d05eba
      Image:         localhost:5000/nettest:local
      Image ID:      localhost:5000/nettest@sha256:63f21b9d8065d60870c2bca980dbc4dc6358edf669661576b8cbee39e3024a2d
      Port:          8080/TCP
      Host Port:     0/TCP
      Command:
        /app/simpleserver
      State:          Waiting
        Reason:       CrashLoopBackOff
      Last State:     Terminated
        Reason:       StartError
        Message:      failed to create containerd task: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "/app/simpleserver": stat /app/simpleserver: no such file or directory: unknown
``` 

Is breaking upgrades currently but not related to this PR 

This should be ready for review, the code is much simpler now due to some ovn-k changes that went in this cycle.

astoycos commented:
Just a heads up @mkolesnik, I'm assuming this ^ is to be expected?

dfarrell07 commented on main.go:
```diff
@@ -252,6 +252,11 @@ func main() {
 	<-stopCh
+
+	if err := cableEngine.Cleanup(); err != nil {
+		klog.Errorf("error cleaning up cableEngine resources before removeing Gateway")
```
```suggestion
		klog.Errorf("error cleaning up cableEngine resources before removing Gateway")
```

dfarrell07 commented on pkg/networkplugin-syncer/handlers/ovn/ovn_logical_router_policy.go:
```diff
@@ -0,0 +1,82 @@
+	// Cleanup any existing lrps not representing the correct set of remote subnets
+	err := libovsdbops.DeleteLogicalRouterPoliciesWithPredicate(ovn.nbdb, ovnClusterRouter, lrpStalePredicate)
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete stale submariner logical route policies")
```
```suggestion
		return errors.Wrapf(err, "failed to delete stale submariner logical router policies")
```

You call these "router policies" everywhere else, finding "route policies" in the logs might not be so clear.

dfarrell07 commented on pkg/networkplugin-syncer/handlers/ovn/ovn_logical_router_policy.go:
```diff
@@ -0,0 +1,82 @@
+
+		if err := libovsdbops.CreateOrUpdateLogicalRouterPolicyWithPredicate(ovn.nbdb,
+			ovnClusterRouter, lrp, lrpSubPredicate); err != nil {
+			return errors.Wrapf(err, "failed to create submariner logical Router policy %v and add it to the ovn cluster router", lrp)
```
```suggestion
			return errors.Wrapf(err, "failed to create submariner logical router policy %v and add it to the ovn cluster router", lrp)
```

Super-nit but while I'm comparing these error messages.

dfarrell07 commented on go.mod:
```diff
@@ -1,30 +1,32 @@
 	k8s.io/client-go v1.5.2
 	k8s.io/klog v1.0.0
-	k8s.io/utils v0.0.0-20210305010621-2afb4311ab10
+	k8s.io/klog/v2 v2.30.0
```
I think this klog bump is breaking CI?

```
#11 8.480 [15:49:43.881] [submariner]$ CGO_ENABLED=0 go build -trimpath -ldflags -s -w -X main.VERSION=devel-1a5ac81 -o bin/linux/amd64/submariner-gateway .
#11 80.41 # sigs.k8s.io/controller-runtime/pkg/log
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/log.go:50:31: cannot use NullLogger{} (value of type NullLogger) as type logr.Logger in argument to NewDelegatingLogger
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/log.go:54:24: cannot use Log (variable of type *DelegatingLogger) as type logr.Logger in variable declaration
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/log.go:56:16: assignment mismatch: 1 variable but logr.FromContext returns 2 values
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/null.go:30:21: cannot use NullLogger{} (value of type NullLogger) as type logr.Logger in variable declaration
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/null.go:48:37: undefined: logr.InfoLogger
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/null.go:54:9: cannot use log (variable of type NullLogger) as type logr.Logger in return statement
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/null.go:59:9: cannot use log (variable of type NullLogger) as type logr.Logger in return statement
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/deleg.go:103:9: cannot use res (variable of type *DelegatingLogger) as type logr.Logger in return statement
#11 80.41 vendor/sigs.k8s.io/controller-runtime/pkg/log/deleg.go:116:9: cannot use res (variable of type *DelegatingLogger) as type logr.Logger in return statement
#11 105.6 make: *** [Makefile:85: bin/linux/amd64/submariner-gateway] Error 2
```

astoycos replied:
Yep Looking at fixing it now :)

tpantelis replied:
yeah we need to keep `logr` at 0.4.0 so we need to keep K8s at 0.21.11 and `klog/v2` at v2.9.0 (or pin it).

astoycos replied:
So the OVN-k Bits require we have k8s 23.3..... Thankfully I was able to bump k8s version (unpinning all pinned versions) without any issues

tpantelis replied:
We ran into issues in **submariner-operator** with **lighthouse** bumped to v0.23.5 - we had to pin a bunch of things. We're going to have to do the same here (ie pin k8s 0.21.11). `sigs.k8s.io/controller-runtime` v0.11.2 uses the backwards incompatible `logr` 1.2.0 - we need to directly set or pin it to `sigs.k8s.io/controller-runtime` v0.9.7. Also `github.com/go-logr/logr` pinned to v0.4.0 and `k8s.io/klog/v2` to v2.9.0.

tpantelis commented on pkg/networkplugin-syncer/handlers/ovn/connection.go:
```diff
@@ -113,3 +113,57 @@ func getOVNTLSConfig(pkFile, certFile, caFile string) (*tls.Config, error) {
 	}, nil
 }
+
+func createLibovsdbClient(dbAddress string, tlsConfig *tls.Config, dbModel model.ClientDBModel) (libovsdbclient.Client, error) {
```
```suggestion
func createLibOvsdbClient(dbAddress string, tlsConfig *tls.Config, dbModel model.ClientDBModel) (libovsdbclient.Client, error) {
```

astoycos replied:
`Libovsdb` is a single word here :)

tpantelis commented on pkg/networkplugin-syncer/handlers/ovn/ovn_submariner_infrastructure.go:
```diff
@@ -0,0 +1,254 @@
+		return err
+	}
+
+	return nil
```
```suggestion
	return ovn.reconcileSubOvnLogicalRouterPolicies(remoteSubnets)
```

tpantelis commented on pkg/networkplugin-syncer/handlers/ovn/ovn_submariner_infrastructure.go:
```diff
@@ -0,0 +1,254 @@
+	submarinerRouter.Options["chassis"] = chassis.Name
+
+	err = libovsdbops.CreateOrUpdateLogicalRouter(ovn.nbdb, submarinerRouter)
+	if err != nil {
+		return errors.Wrap(err, "failed to set chassis option on ovn submariner router")
+	}
+
+	return nil
```
```suggestion
	return errors.Wrap(err, "failed to set chassis option on ovn submariner router")
```

astoycos replied:
Wouldn't this result in us always returning an error? And in this case we need to ensure that the `libovsdbops` function call  is wrapped here so that if it's encountered the user can understand the context

tpantelis replied:
No, `errors.Wrap` et al conveniently check if `err` is nil.

astoycos replied:
Ah Nice I didn't know that :)

astoycos replied:
Done

tpantelis commented on pkg/networkplugin-syncer/handlers/ovn/ovn_submariner_infrastructure.go:
```diff
@@ -0,0 +1,254 @@
+	}
+
+	err = libovsdbops.CreateOrUpdateLogicalRouterPorts(ovn.nbdb, &subLogicalRouter, &subRouterTosubGatewayLrp)
+	if err != nil {
+		return errors.Wrap(err, "failed to add gateway to submariner router LSP to the submariner gateway switch")
+	}
+
+	return nil
```
```suggestion
	return errors.Wrap(err, "failed to add gateway to submariner router LSP to the submariner gateway switch")
```

astoycos replied:
Same comment as above ^

astoycos replied:
Done

tpantelis commented on pkg/networkplugin-syncer/handlers/ovn/ovn_submariner_infrastructure.go:
```diff
@@ -0,0 +1,254 @@
+	klog.V(log.DEBUG).Infof("reconciling north static routes on %q router for subnets %v", submarinerLogicalRouter,
+		remoteSubnets.Elements())
+
+	err := ovn.reconcileSubOvnLogicalRouterStaticRoutes(submarinerUpstreamRPort, HostUpstreamIP, remoteSubnets)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
```suggestion
	return ovn.reconcileSubOvnLogicalRouterStaticRoutes(submarinerUpstreamRPort, HostUpstreamIP, remoteSubnets)
```

tpantelis commented on pkg/networkplugin-syncer/handlers/ovn/ovn_submariner_infrastructure.go:
```diff
@@ -0,0 +1,254 @@
+	klog.V(log.DEBUG).Infof("reconciling south static routes on %q for subnets %v", submarinerLogicalRouter,
+		localSubnets)
+
+	err := ovn.reconcileSubOvnLogicalRouterStaticRoutes(submarinerDownstreamRPort, ovnClusterSubmarinerIP, localSubnets)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
```suggestion
	return ovn.reconcileSubOvnLogicalRouterStaticRoutes(submarinerDownstreamRPort, ovnClusterSubmarinerIP, localSubnets)
```

astoycos committed:
Fix Unit tests after bumping to K8s 1.23

For a Fake k8s client we must now
explicitly register all resouces
we expect to list in the scheme.

Otherwise we'll run into
https://github.com/kubernetes/client-go/issues/914

Signed-off-by: Andrew Stoycos <astoycos@redhat.com>

astoycos commented:
I still need to work on vendoring issues, since now they're breaking unit tests

tpantelis commented on pkg/globalnet/controllers/controllers_suite_test.go:
```diff
@@ -114,6 +116,10 @@ func newTestDriverBase() *testDriverBase {
+	// TODO: Remove this workaround for https://github.com/kubernetes/client-go/issues/949 once
+	// admiral has been updated
+	t.scheme.AddKnownTypeWithName(schema.GroupVersionKind{Group: "fake-dynamic-client-group", Version: "v1", Kind: "List"}, &unstructured.UnstructuredList{})
+
```
I think we should pin to K8s 0.21.11 to avoid having to do this. The rest of the submariner projects use 0.21.11 so I think we should be consistent.

astoycos replied:
@tpantelis  I was more just trying to figure out why that was happening :) ..... Also I believe [this problem](https://github.com/kubernetes/client-go/issues/949) occurred from 0.19 -> 0.20 so we'll probably still run into this specific issue 

> We ran into issues in submariner-operator with lighthouse bumped to v0.23.5 - we had to pin a bunch of things. We're going to have to do the same here (ie pin k8s 0.21.11). sigs.k8s.io/controller-runtime v0.11.2 uses the backwards incompatible logr 1.2.0 - we need to directly set or pin it to sigs.k8s.io/controller-runtime v0.9.7. Also github.com/go-logr/logr pinned to v0.4.0 and k8s.io/klog/v2 to v2.9.0.

I'll give pinning to 0.21.11 a try...But currently there's some dependencies with the imported OVN-K libraries which are forcing the version up to v0.23.3 since that's what's defined here -> https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/go.mod#L38.  Also something to keep in mind... We don't use `logr` in this repo so I don't think that's a problem anymore... I updated klog/v2 to v2.30 (in this current iteration) and it looks like everything is running alright 

But I do need to ensure that we can import those ovn-k libraries without depending on k8s version.. Or else then every time ovn-k bumps their K8s version we'll have to, creating an unnecessary dependency (unless we are always just keeping it pinned...)

astoycos replied:
For example when I pin to 0.21.11 

```
#14 29.60 # github.com/ovn-org/ovn-kubernetes/go-controller/pkg/libovsdbops
#14 29.60 vendor/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/libovsdbops/transact.go:21:20: undefined: wait.PollImmediateUntilWithContext
```

tpantelis replied:
> Also something to keep in mind... We don't use `logr` in this repo so that's not a problem anymore... 

But I intend/want to use `logr`, specifically for zerolog.

> But I do need to ensure that we can import those ovn-k libraries without depending on k8s version.. Or else then every time ovn-k bumps their K8s version we'll have to, creating an unnecessary dependency.

I see that you're using the `master` branch from `github.com/ovn-org/ovn-kubernetes`. Not sure how stable that is but I guess we have to since no prior version has a `go.mod` file.

tpantelis replied:
`wait.PollImmediateUntilWithContext ` was added in K8s 0.22 so perhaps try pinning to 0.22.9, which uses `k8s.io/klog/v2` v2.9.0 and `logr` v0.4.0.

astoycos replied:
> I see that you're using the master branch from github.com/ovn-org/ovn-kubernetes. Not sure how stable that is but I guess we have to since no prior version has a go.mod file.

Right they have no upstream release versioning process so it is a bit unstable.... granted these bits were already pretty unstable, and regardless I don't think what we have here is necessarily the best long-term ovn-k integration solution since any way we cut it there will always be a tightly coupled dependency with the ovn-k version

tpantelis replied:
Ideally we would bump to K8s 0.23 across the board. At one point **lighthouse** was using 0.23 due to `coredns` bump but we downgraded due to **submariner-operator** issues.  Obviously **submariner** is building fine with 0.23 with this PR.  I'm not sure what's holding us back. @skitt? Is it the older operator framework in **submariner-operator**?

tpantelis replied:
I tried pinning to K8s 0.22.9 et al but hit a showstopper:

```
github.com/submariner-io/submariner/pkg/networkplugin-syncer/handlers/ovn imports
        github.com/ovn-org/libovsdb/client imports
        github.com/go-logr/stdr imports
        github.com/go-logr/logr/funcr: no required module provides package github.com/go-logr/logr/funcr; to add it:
        go get github.com/go-logr/logr/funcr
~/go/src/github.com/submariner-io/submariner>go get github.com/go-logr/logr/funcr
go: module github.com/go-logr/logr@upgrade found (v1.2.3, replaced by github.com/go-logr/logr@v0.4.0), but does not contain package github.com/go-logr/logr/funcr
```

So `github.com/go-logr/logr/funcr` was added in `logr` v1.0.0 which also introduces the backwards incompatible `Logger` change from interface to struct. So in order to use `libovsdb`, we need K8s 0.23 et al. We'll see what happens when the **submariner** version is bumped in **submariner-operator** 😄 Of course bumping everything to K8s 0.23 would help.

astoycos replied:
Yeah I experienced the exact same thing :(  

And sounds good, I'll let this sit for now since CI is green and continue to monitor the best way forward

tpantelis replied:
Discussed with @skitt - after bumping the operator framework version and adjusting our operator code (planned epic for this release) and requiring go 1.17, there should be no obstacle in bumping to K8s 0.23 across the projects.

dfarrell07 reviewed (approved):
LGTM! @sridhargaddam / @vthapar / @aswinsuryan does anyone else want to review or can we merge this?

sridhargaddam commented:
> LGTM! @sridhargaddam / @vthapar / @aswinsuryan does anyone else want to review or can we merge this?

@dfarrell07 I could not look into this as I was occupied with other stuff, but @vthapar wanted to try this out on an OCP cluster. I'll let him comment. I'm fine if we want to go ahead and merge it.

vthapar commented:
@dfarrell07 @sridhargaddam While testing I ran into a issue with OCP 4.10. This fix requires OCP 4.11 or later to work. I've been unable to get a working setup with OCP 4.11 I'd like to confirm it working with 4.11 at least once before we merge it. I should have results in a day at most.

vthapar reviewed (approved):
I didn't review the code but did test it out and confirmed E2E to be passing in both kind and AWS+OCP.
